### PR TITLE
Add a dedicated `correctors.metrics` module for under- and overfitting metrics

### PR DESCRIPTION
**Pull request by barentsen**
Over in #855, @jcsmithhere opened a fabulous PR to add a much-improved `CBVCorrector` class to Lightkurve.

One of the questions raised in that PR was whether it would be good to move the generic over- and under-fitting metrics it implements outside of `CBVCorrector`.  This seems like an excellent idea because they are sufficiently generic that any correction method could benefit from them.

Because I struggle to think about software design without trying things out, I went ahead and created this branch which explores how best to add Jeff's over- and underfitting metrics into the existing Lightkurve infrastructure.

Specifically, this PR makes the following changes:
* I added a dedicated `correctors.metrics` module to host the `under_fitting_metric` and `over_fitting_metric` methods created by @jcsmithhere in #855 as two standalone functions.
* To make it easier to add alternative under/over-fitting metrics in the future, I went ahead and gave these functions more specific names, namely `overfit_metric_lombscargle` and `underfit_metric_neighbors`.  I don't necessarily love these names, so I'm happy for them to be changed!
* I added two simple unit tests focused specifically on these metrics in `correctors/tests/test_metrics.py`.
* I added `compute_underfit_metric()` and `compute_overfit_metric()` to the base `Corrector` class over in `correctors/corrector.py`.  These methods are lightweight wrappers for the functions defined in `correctors/metrics.py` and makes them easily accessible to *any* corrector class.
* I expanded the base `Corrector` class over in `correctors/corrector.py` to better document a vision on a uniform structure for all corrector classes.

Some minor changes:
* I ran `black` on `metrics.py` for consistent formatting. (I plan to run `black` on all of Lightkurve in the near future.)
* I changed a couple of `print()` statements in `under_fitting_metric` to use `log.info()`, so they can be turned on or off (using e.g. `lk.log.setLevel("INFO").`).  Admittedly the under fitting metric can take a long time to run, so maybe we do want to add a `verbose=True` argument which would enable a `tqdm` progress bar to be displayed while downloading the neighbors.
* I noticed that the `under_fitting_metric` function implemented logic to search and obtain neighbor light curves.  Because this is a really neat feature, I moved that logic to a dedicated `LightCurve.search_neighbors()` method.
* The `under_fitting_metric` function was keeping a local cache (`self.lc_neighborhood`).  For simplicity, I modified `lightkurve/search.py` to cache the search operations directly, so client functions no longer need to worry about caching search operations. (File downloads were already being cached.)
* `under_fitting_metric` used a dynamic search radius to identify a sufficient number of neighbors.  I was a bit lazy during refactoring and wondered if we can simply use a very large default radius instead, but I haven't verified that this will work well in most cases.

If we merge this PR (with @jcsmithhere's approval), I think we'll be very close to merging the rest of `CBVCorrector` via #855.  The next steps would be:
1. Update #855 to take advantage of the fact that the metrics have been merged into Lightkurve via this PR.
2. Have a think about remaining design questions, including:
   * How best to support the use case demonstrated in the "More Generalized Design Matrix Priors" section of the excellent tutorial notebook in #855 (in which additional data is added to the CBV Design Matrix).
   * How the optimization and scan plot features of `CBVCorrector` may be adopted by other correctors in a consistent way.

Thanks again for the excellent work @jcsmithhere.  These metrics are a really neat contribution!  Please feel free to push commits directly to this PR branch, should you wish to do so.

### CODE REVIEW

**Comment by jcsmithhere** [comment]
Hi @barentsen! Sorry, I'm just now getting to this. I was a bit busy these past couple weeks. I'll look through this PR. I think it is a good idea these goodness metrics be available outside the `cbvCorrector` so I'll try to get this all to work.

**Review by jcsmithhere — COMMENTED** [review]
Hi @barentsen, Thanks for putting this together. Overall, I think this can work. I just have a couple comments, however, that might change once I begin really testing it with `CBVCorrector`. 

1. There's some odd grammar with using `neighbor` as a noun instead of an adjective (`neighboring`). Was that intentional?
3. The goodness metrics are supposed to be really fast to run inside a Loss Function for an optimizer. Therefore I would like to store the aligned or interpolated neighboring flux, instead of recomputing them every time the metric is run. 
3. Regarding the default search radius for the neighborhood, I think a default in the generic search function at 3600 arcseconds is probably fine. But it might not be that much work to have a different default for each mission depending on the plate scale. If the lightcurve object does not have a Kepler or TESS mission associated with it then use the generic default. 

The comments below are just that, comments. I can make the changes. I'll push my changes to this branch. I really won't know how well this works until I test it with my `CBVCorrector` but I think it should work as-is (except for the speed issue).

File: lightkurve/correctors/metrics.py
Comment (jcsmithhere) [inline lightkurve/correctors/metrics.py:172]: `_download_neighbors` is really slow to run.  I haven't completely dived through the code yet so am not certain if it is using the cache. But even if it is, there's a lot of functionality in there being called every time `underfit_metric_neighbors` is called and probably not viable in an optimizer loop to be run thousands of times. I think we need to have some type of cache for `underfit_metric_neighbors` so that `lc.search_neighbors` is only called once.

**Review by jcsmithhere — CHANGES_REQUESTED** [review]
I think we need to reconsider how this is being done. `_download_neighbors` is really slow and too slow for an optimizer. I think we need some type of local cache for the underfitting metric for speed.

File: lightkurve/correctors/metrics.py
Comment (jcsmithhere) [reply lightkurve/correctors/metrics.py:172]: I'm thinking the thing to do is inside the generic `Corrector` class have a cache, we then tell overfit_metric_lombscargle to either search for the neighborhood targets, or use a cache inside the `Corrector` object. This would be similar to what I had inside `CBVCorrector`,  just easier for non-cbvcorrector users to use.

File: lightkurve/correctors/metrics.py
Comment (barentsen) [reply lightkurve/correctors/metrics.py:306]: We can use `neighboring` instead of `neighbor` -- this was really just poor grammar on my part.

File: lightkurve/search.py
Comment (barentsen) [reply lightkurve/search.py]: (The `SearchResult.distance` property has been added via #913 in meanwhile, with correct comments, so this particular issue has gone away now that I rebased on master.)

File: lightkurve/correctors/tests/test_metrics.py
Comment (barentsen) [reply lightkurve/correctors/tests/test_metrics.py:49]: This test does pass, so at least the edge cases behave as expected, which is great!

File: lightkurve/correctors/metrics.py
Comment (barentsen) [reply lightkurve/correctors/metrics.py:172]: Thanks Jeff!  I confirm that `_download_neighbors` is very slow to run, not just on the first call, but also on every subsequent call.  That is unexpected.  It looks like the `download_all()` instruction isn't using the files cached on disk. My fault for not verifying that the caching is working correctly.  I'm going to investigate this bug.

Even when the caching of `download_all()` is fixed, I agree we will want to cache some of the additional alignment/interpolation work that takes place inside `underfit_metric_neighbors()`.  The cleanest way to do this will be to move the work that can be cached to a separate function (e.g. `_download_and_align_neighbors()`) and decorate that function with the `@memoization.cached` decorator which will automatically take care of caching for us.

File: lightkurve/correctors/metrics.py
Comment (barentsen) [reply lightkurve/correctors/metrics.py:199]: Agreed!  See related response below.

File: lightkurve/correctors/metrics.py
Comment (barentsen) [reply lightkurve/correctors/metrics.py:50]: Correct.  I want to prevent the concept of the `Corrector.cadence_mask` property from leaking outside of the `Corrector` class. (where `cadence_mask` is used to flag cadences that ought to be ignored while fitting the noise model).

When calling the standalone metric functions, a user can apply a mask to the input light curves as necessary, i.e. one can do `overfit_metric(original_lc[mask], corrected_lc[mask])`, so no need for a separate mask keyword there.

Comment (barentsen) [reply lightkurve/correctors/metrics.py:50]: (But we do indeed need to remove the outdated comment on line 47!)

**Comment by barentsen** [comment]
Quick update:
* I agree with all the excellent comments made by @jcsmithhere above.
* The comments revealed that `_download_neighbors()` isn't using the MAST download cache.  I have yet to identify the cause.
* I rebased on master tonight to resolve two merge conflicts which don't otherwise affect anything discussed above.

**Comment by barentsen** [comment]
Here is the current line-by-line execution cost of `underfit_metric_neighbors`, confirming that the majority of the cost can be cached:

```python
>>> import lightkurve as lk 
>>> from lightkurve.correctors.metrics import underfit_metric_neighbors
>>> lc = lk.search_lightcurve("Pi Men")[0].download()
>>> %load_ext line_profiler  # install using `pip install line_profiler`
>>> %lprun -f underfit_metric_neighbors underfit_metric_neighbors(lc)
Timer unit: 1e-06 s

Total time: 83.0374 s
File: /Users/gb/dev/lightkurve/lightkurve/correctors/metrics.py
Function: underfit_metric_neighbors at line 117

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   117                                           def underfit_metric_neighbors(
   118                                               corrected_lc: LightCurve,
   119                                               radius: float = 6000,
   120                                               min_targets: int = 30,
   121                                               max_targets: int = 50,
   122                                               interpolate: bool = False,
   123                                           ):
   124                                               """This goodness metric measures the degree of under-fitting of the
   125                                               CBVs to the light curve. It does so by measuring the mean residual target to
   126                                               target Pearson correlation between the target under study and a selection of
   127                                               neighboring SPOC SAP target light curves.
   128                                           
   129                                               This function will begin with the given search_radius in arcseconds and
   130                                               finds all neighboring targets. If not enough were found (< min_targets)
   131                                               the search_radius is increased until a minimum number are found.
   132                                           
   133                                               For TESS, the default search_radius is 5000 arcseconds.
   134                                               For Kepler/K2, the default search_radius is 1000 arcseconds
   135                                           
   136                                               The downloaded neighboring targets will normally be "aligned" to the
   137                                               corrected_lc meaning the cadence numbers are used to align the targets
   138                                               to the corrected_lc. However, if the CBVCorrector object was
   139                                               instantiated with interpolated_cbvs=True then the targets will be
   140                                               interpolated to the corrected_lc cadence times.
   141                                           
   142                                               The returned under-fitting goodness metric is callibrated such that a
   143                                               value of 0.95 means the residual correlations in the target is
   144                                               equivalent to chance correlations of White Gaussian Noise.
   145                                           
   146                                               Parameters
   147                                               ----------
   148                                               corrected_lc : LightCurve
   149                                                   Light curve from which systematics have been removed.
   150                                               radius : float
   151                                                   Search radius to find neighboring targets in arcseconds
   152                                               min_targets : int
   153                                                   Minimum number of targets to use in correlation metric
   154                                                   Using too few can cause unreliable results. Default = 30
   155                                               max_targets : int
   156                                                   Maximum number of targets to use in correlation metric
   157                                                   Using too many can slow down the metric due to large data
   158                                                   download. Default = 50
   159                                           
   160                                               Returns
   161                                               -------
   162                                               under_fitting_metric : float
   163                                                   A float in the range [0,1] where 0 => Bad, 1 => Good
   164                                               """
   165                                               # Download neighbor light curves
   166         2   73538179.0 36769089.5     88.6      lcfCol = _download_neighbors(
   167         1          5.0      5.0      0.0          lc=corrected_lc,
   168         1          3.0      3.0      0.0          radius=radius,
   169         1          3.0      3.0      0.0          min_targets=min_targets,
   170         1          3.0      3.0      0.0          max_targets=max_targets,
   171         1          3.0      3.0      0.0          flux_column="sap_flux",
   172                                               )
   173                                           
   174                                               # Pre-process the neighbor light curves
   175         1          4.0      4.0      0.0      lc_neighborhood = []
   176         1          2.0      2.0      0.0      lc_neighborhood_flux = []
   177                                               # Extract SAP light curves
   178                                               # We want zero-centered median normalized light curves
   179        51        617.0     12.1      0.0      for lc in lcfCol:
   180        50    5446265.0 108925.3      6.6          lcSAP = lc.remove_nans().normalize()
   181        50      25099.0    502.0      0.0          lcSAP.flux -= 1.0
   182        50        156.0      3.1      0.0          lc_neighborhood.append(lcSAP)
   183                                                   # Align or interpolate the neighboring target with the target under study
   184        50         82.0      1.6      0.0          if interpolate:
   185                                                       # Interpolate to corrected_lc cadence times
   186                                                       fInterp = PchipInterpolator(
   187                                                           lc_neighborhood[-1].time.value,
   188                                                           lc_neighborhood[-1].flux.value,
   189                                                           extrapolate=False,
   190                                                       )
   191                                                       lc_neighborhood_flux.append(fInterp(corrected_lc.time.value))
   192                                                   else:
   193                                                       # The CBVs were aligned so also align the neighboring
   194                                                       # lightcurves
   195       100      56938.0    569.4      0.1              lc_trim_mask = np.in1d(
   196        50       1318.0     26.4      0.0                  lc_neighborhood[-1].cadenceno, corrected_lc.cadenceno
   197                                                       )
   198        50    3788224.0  75764.5      4.6              lc_neighborhood_flux.append(lc_neighborhood[-1][lc_trim_mask].flux.value)
   199                                           
   200                                               # Store the unmolested lightcurve neighborhood but also save the
   201                                               # aligned or interpolated neighborhood flux
   202         1         19.0     19.0      0.0      from .. import LightCurveCollection
   203                                           
   204         1         23.0     23.0      0.0      lc_neighborhood = LightCurveCollection(lc_neighborhood)
   205         1          2.0      2.0      0.0      lc_neighborhood_flux = lc_neighborhood_flux
   206                                           
   207                                               # If there happens to be any cadences in the corrected_lc
   208                                               # that are not in the neighboring targets then those need to
   209                                               # be NaNed.
   210                                               # If we interpolated the CBVs then this should not occur
   211         1     112607.0 112607.0      0.1      corrected_lc = corrected_lc.copy().remove_nans()
   212         1          3.0      3.0      0.0      if not interpolate:
   213         2         15.0      7.5      0.0          corrected_lc_trim_mask = np.logical_not(
   214         1       1574.0   1574.0      0.0              np.in1d(corrected_lc.cadenceno, lc_neighborhood[0].cadenceno)
   215                                                   )
   216         1         38.0     38.0      0.0          corrected_lc.flux.value[corrected_lc_trim_mask] = np.nan
   217                                           
   218                                               # Create fluxMatrix. The last entry is the target under study
   219         1         24.0     24.0      0.0      fluxMatrix = np.zeros((len(lc_neighborhood_flux[0]), len(lc_neighborhood_flux) + 1))
   220        51        150.0      2.9      0.0      for idx in np.arange(len(fluxMatrix[0, :]) - 1):
   221        50       8866.0    177.3      0.0          fluxMatrix[:, idx] = lc_neighborhood_flux[idx]
   222                                               # Add in target under study, and median normalize it
   223         1      34990.0  34990.0      0.0      fluxMatrix[:, -1] = corrected_lc.normalize().flux
   224         1        250.0    250.0      0.0      fluxMatrix[:, -1] -= 1.0
   225                                           
   226                                               # Ignore NaNs
   227         1        142.0    142.0      0.0      mask = ~np.isnan(corrected_lc.flux)
   228         1       5233.0   5233.0      0.0      fluxMatrix = fluxMatrix[mask, :]
   229                                           
   230                                               # Determine the target-target correlation between target and
   231                                               # neighborhood
   232         1      15905.0  15905.0      0.0      correlationMatrix = _compute_correlation(fluxMatrix)
   233                                           
   234                                               # The selection basis for targets used for the PDC-MAP SVD  uses median
   235                                               # absolute correlation per star.  However, here we wish to overemphasize
   236                                               # any residual correlation between a handfull of targets and not the
   237                                               # overall correlation (which should almost always be low).
   238                                           
   239                                               # We want a residual correlation larger than random correlations of WGN
   240                                               # to mean a meaningful correlation. The median Pearson correlation of
   241                                               # WGN of nCadences is approximated by the equation:
   242                                               # 0.0010288 + 0.80304 nCadences^ -0.50128
   243         1         10.0     10.0      0.0      nCadences = len(fluxMatrix[:, 0])
   244         1          3.0      3.0      0.0      beta = [0.0007, 0.8083, -0.5023]
   245         1          6.0      6.0      0.0      WGNCorrelation = beta[0] + beta[1] * (nCadences ** (beta[2]))
   246                                           
   247                                               # badLimit is the goodness value for WGN correlations
   248                                               # I.e. anything above this goodness value is equivalent to random correlations
   249                                               # I.e. 0.95 = sigmoidInv(WGNCorr * correlationScale)
   250         1          2.0      2.0      0.0      badLimit = 0.95
   251         1         15.0     15.0      0.0      correlationScale = 1 / (WGNCorrelation) * np.log((2.0 / badLimit) - 1.0)
   252                                           
   253                                               # Over-emphasize any individual correlation groups. Note the power of
   254                                               # three after taking the absolute value
   255                                               # of the correlation. Also, the mean is used so that outliers are *not* ignored.
   256                                               # Zero diagonal elements
   257         2        218.0    109.0      0.0      correlationMatrix = np.tril(correlationMatrix, k=-1) + np.triu(
   258         1          3.0      3.0      0.0          correlationMatrix, k=+1
   259                                               )
   260                                           
   261                                               # Add up the correlation over all targets ignoring NaNs (no corrected fit)
   262         1        344.0    344.0      0.0      correlation = correlationScale * np.nanmean(np.abs(correlationMatrix) ** 3, axis=0)
   263                                           
   264                                               # We only want the last entry, which is for the target under study
   265         1          3.0      3.0      0.0      correlation = correlation[-1]
   266                                           
   267                                               # We want the goodness to span (0,1]
   268                                               # Use twice a reversed sigmoid to get a [0,1] range mapped from a [0,inf) range
   269         1          7.0      7.0      0.0      def sigmoidInv(x):
   270                                                   return 2.0 / (1 + np.exp(x))
   271                                           
   272         1         12.0     12.0      0.0      metric = sigmoidInv(correlation)
   273                                           
   274         1          2.0      2.0      0.0      return metric
```

**Comment by barentsen** [comment]
I pushed a new commit (6c1168b) which replaces the `_download_neighbors` helper function with a more comprehensive function called `_download_and_preprocess_neighbors` which is cached using the `memoization.cached` feature.

While the first call to `underfit_metric_neighbors` still takes a long time (>20s), the second and all subsequent calls now execute in ~200ms:

```python
>>> import lightkurve as lk
>>> from lightkurve.correctors.metrics import underfit_metric_neighbors
>>> lc = lk.search_lightcurve("Pi Men")[0].download()
>>> %time underfit_metric_neighbors(lc)
CPU times: user 22.4 s, sys: 1.03 s, total: 23.4 s
Wall time: 1min 3s
>>> %time underfit_metric_neighbors(lc)
CPU times: user 140 ms, sys: 29.5 ms, total: 170 ms
Wall time: 211 ms
```

It is worth noting that the caching mechanism depends on a custom key that will determine whether a previous result will be re-used, cf.

https://github.com/barentsen/lightkurve/blob/6c1168bb0c2ab0cc99eaf38a623502e12c361143/lightkurve/correctors/metrics.py#L264-L265

I defined this key to be a function of `corrected_lc.ra`, `corrected_lc.dec`, and `corrected_lc.cadenceno`, which I think are the only properties of `corrected_lc` which would invalidate the cache if they changed.

(A related performance issue is that `search.download_all()` is surprisingly slow even when the light curve files have been cached on disk.  I am going to investigate and solve this via the separate issue #914 because it will only affect our first call to `underfit_metric_neighbors` now.)

File: lightkurve/correctors/metrics.py
Comment (barentsen) [reply lightkurve/correctors/metrics.py]: I went ahead and replaced every instance of "neighbor light curve" with "neighboring light curve".

File: lightkurve/correctors/metrics.py
Comment (barentsen) [reply lightkurve/correctors/metrics.py:51]: (I went ahead and removed the outdated comment on line 47 which read `# Also ignore masked cadence`.)

File: lightkurve/correctors/metrics.py
Comment (jcsmithhere) [reply lightkurve/correctors/metrics.py]: Confirmed that cache is working!

File: lightkurve/correctors/tests/test_metrics.py
Comment (jcsmithhere) [reply lightkurve/correctors/tests/test_metrics.py:49]: I'll leave writing better units tests for a later PR. I don't want to hold up this PR.

File: lightkurve/correctors/metrics.py
Comment (jcsmithhere) [reply lightkurve/correctors/metrics.py:178]: Confirmed caching code is functional and very fast.

Comment (barentsen) [reply lightkurve/correctors/metrics.py:178]: Awesome! 🎉 

Note: I merged #915 into master earlier today, which will speed up the initial call to `download_all()` before the caching provided within `metrics.py` kicks in.

File: lightkurve/correctors/tests/test_metrics.py
Comment (barentsen) [reply lightkurve/correctors/tests/test_metrics.py:49]: :+1:

**Comment by jcsmithhere** [comment]
It all looks good to me, @barentsen!

**Comment by barentsen** [comment]
Thanks @jcsmithhere!!  I rebased on master to resolve the merge conflicts (which were trivial) and will merge once the tests are all green! :+1: :+1: :+1:

**Comment by barentsen** [comment]
Merged into master! 🥳

**Pull request merged**
